cf/mixin/fielddomain.py

```diff
@@ -2408,6 +2408,110 @@ def set_construct(
+                identifiers, with or without the ``'key%'`` prefix.
+
+                {{value match}}
+
+                {{displayed identity}}
+
+            default: optional
+                Return the value of the *default* parameter if the
+                data axes have not been set.
+
+                {{default Exception}}
+
+            {{filter_kwargs: optional}}
+
+                .. versionadded:: NEXTVERSION
+
+        :Returns:
+
+                The removed metadata construct.
+
+        **Examples**
+
+        >>> f = {{package}}.example_field(0)
+        >>> print(f)
+        Field: specific_humidity (ncvar%q)
+        ----------------------------------
+        Data            : specific_humidity(latitude(5), longitude(8)) 1
+        Cell methods    : area: mean
+        Dimension coords: latitude(5) = [-75.0, ..., 75.0] degrees_north
+                        : longitude(8) = [22.5, ..., 337.5] degrees_east
+                        : time(1) = [2019-01-01 00:00:00]
+        >>> f.del_construct('time')
+        <{{repr}}DimensionCoordinate: time(1) days since 2018-12-01 >
+        >>> f.del_construct('time')
+        Traceback (most recent call last):
+            ...
+        ValueError: Can't find unique construct to remove
+        >>> f.del_construct('time', default='No time')
+        'No time'
+        >>> f.del_construct('dimensioncoordinate1')
+        <{{repr}}DimensionCoordinate: longitude(8) degrees_east>
+        >>> print(f)
+        Field: specific_humidity (ncvar%q)
+        ----------------------------------
+        Data            : specific_humidity(latitude(5), ncdim%lon(8)) 1
+        Cell methods    : area: mean
+        Dimension coords: latitude(5) = [-75.0, ..., 75.0] degrees_north
+
+        """
+        # Need to re-define to overload this method since cfdm doesn't
+        # have the concept of cyclic axes, so have to update the
+        # register of cyclic axes when we delete a construct in cf.
+
+        # Get the relevant key first because it will be lost upon deletion.
+        key = self.construct_key(*identity, default=None, **filter_kwargs)
+        # Copy since should never change value of _cyclic attribute in-place
+        cyclic_axes = self._cyclic.copy()
+
+        deld_construct = super().del_construct(
+            *identity, default=default, **filter_kwargs
+        )
+
+        # If the construct deleted was a cyclic axes, remove it from the set
+        # of stored cyclic axes, to update that appropriately. Do this
+        # afterwards because the deletion might not be successful and don't
+        # want to update the cyclic() set unless we know the deletion occurred.
+        if key in cyclic_axes:
+            # The below is to test that the construct was successfully deleted
+            if isinstance(default, Exception) or (
+                not isinstance(default, Exception)
+                and deld_construct != default
+            ):
```

Comment:
FYI (reviewing), I considered using `has_construct` to test whether the deletion occurred successfully, but thought it would only test the construct was not there which may have the potential to hide future bugs, whereas this logic will test that the `super().del_construct()` call directly above did actually manage to delete a construct that was previously there. I do accept it isn't ideal because it is a bit of a clunky way to check. It left me wishing the signature return included a Boolean to confirm whether or not the deletion happened, explicitly, but I think that would complicate the signature and output too much.

Not sure what you think, David - is this best or should I use `has_construct` or some other means, to check whether the `del_construct` operation above worked?

Reply:
Hi Sadie, we usually deal with this by setting the default to `None` and then handling that easily testable case (see above)